spinnaker/spinnaker.go

```diff
@@ -479,12 +479,35 @@ func (s Spinnaker) asgs(appName, account, clusterName string) (result []spinnake
 	return asgs, nil
 }
 
-// CloudProvider returns the cloud provider for a given account
-func (s Spinnaker) CloudProvider(account string) (provider string, err error) {
-	url := s.accountURL(account)
+// CloudProvider returns the cloud provider for a given account name
+func (s Spinnaker) CloudProvider(name string) (provider string, err error) {
+	account, err := s.account(name)
+	if err != nil {
+		return "", err
+	}
+
+	if account.CloudProvider == "" {
+		return "", errors.New("no cloudProvider field in response body")
+	}
+
+	return account.CloudProvider, nil
+}
+
+// account represents a spinnaker account
+type account struct {
+	CloudProvider string `json:"cloudProvider"`
+	Name          string `json:"name"`
+	Error         string `json:"error"`
+}
+
+// account returns an account by its name
+func (s Spinnaker) account(name string) (*account, error) {
```

Comment:
Are you plan to update the `account` returned here? Pointers serve one purpose, to share a value with a function so the function can read and write to that value even though the value does not exist directly inside its own frame
If you don't need to share - prefer to see value semantics. Each time when function receive or return pointers - this causes memory allocation on heap.

Reply:
Understood. Here, `account` is a pointer because of the `nil` case : when there's an error, I return no `account` value. I must admit that I have never proceed otherwise.

Is this a good solution ? Are there any best practices for these throwable vars ?

```diff
// account returns an account by its name
func (s Spinnaker) account(name string) (account, error) {
	url := s.accountsURL(true)
	resp, err := s.client.Get(url)
+	var ac account

	// Usual HTTP checks
	if err != nil {
+		return ac, errors.Wrapf(err, "http get failed at %s", url)
	}

	defer func() {
		if cerr := resp.Body.Close(); cerr != nil && err == nil {
			err = errors.Wrap(err, fmt.Sprintf("body close failed at %s", url))
		}
	}()

	body, err := ioutil.ReadAll(resp.Body)
	if err != nil {
+		return ac, errors.Wrapf(err, "body read failed at %s", url)
	}

	var accounts []account
	err = json.Unmarshal(body, &accounts)
	if err != nil {
+		return ac, errors.Wrap(err, "json unmarshal failed")
	}
	statusKO := resp.StatusCode != http.StatusOK

	// Finally find account
	for _, a := range accounts {
		if a.Name != name {
			continue
		}
		if statusKO {
			if a.Error == "" {
+				return ac, errors.Errorf("unexpected status code: %d. body: %s", resp.StatusCode, body)
			}

+			return ac, errors.Errorf("unexpected status code: %d. error: %s", resp.StatusCode, a.Error)
		}

		return a, nil
	}

+	return ac, errors.New("the account name doesn't exist")
}
```

Reply:
I am fine with either but looking at the other parts of the code, it seems like the value semantics are indeed preferred for those cases where type is used locally to just pass a group of values between a couple of methods.

I would not worry about performance or heap-optimizations